### ISSUE TITLE
fieldManager errors can fail create/update/non-apply-patch API calls

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#90904]
**What happened**:

FieldManager errors were encountered in an Update API request (xref https://github.com/openfaas/faas-netes/issues/630#issuecomment-625440515):
```
Unexpected status: 500, message: failed to update object (Update for openfaas.com/v1, Kind=Function) managed fields: failed to convert live object to proper version: /, Kind= is unstructured and is not suitable for converting to "openfaas.com/v1"
```

This error blocked an update request that worked in prior Kubernetes versions.

**What you expected to happen**:

The particular error that occurred there should be tracked down and resolved separately, but I would expect field manager errors only block apply requests, not create/update/non-apply-patch requests.

In these non-apply handler locations, fieldManager errors are a "should not happen" type of failure. If a fieldManager-related error is encountered, I'd expect to log the error (and perhaps bump a server-side-apply-error metric), but to then recover and allow the create/update/patch request to continue (potentially by resetting the managedFields to avoid persisting bad data):

```
grep "Manager.Update(" staging/src/k8s.io/apiserver/pkg/endpoints/handlers/*.go
staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:				obj, err = scope.FieldManager.Update(liveObj, obj, managerOrUserAgent(options.FieldManager, req.UserAgent()))
staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:		if objToUpdate, err = p.fieldManager.Update(currentObject, objToUpdate, managerOrUserAgent(p.options.FieldManager, p.userAgent)); err != nil {
staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:		if newObj, err = p.fieldManager.Update(currentObject, newObj, managerOrUserAgent(p.options.FieldManager, p.userAgent)); err != nil {
staging/src/k8s.io/apiserver/pkg/endpoints/handlers/update.go:					obj, err := scope.FieldManager.Update(liveObj, newObj, managerOrUserAgent(options.FieldManager, req.UserAgent()))
```

/priority critical-urgent
/sig api-machinery
/assign @apelisse

**Comment by lavalamp** [comment]
> In these non-apply handler locations, fieldManager errors are a "should not happen" type of failure. If a fieldManager-related error is encountered, I'd expect to log the error (and perhaps bump a server-side-apply-error metric), but to then recover and allow the create/update/patch request to continue (potentially by resetting the managedFields to avoid persisting bad data):

This escalates transient errors (e.g., failure to call conversion webhooks) into permanent data loss. That doesn't seem right.

I think we should actually fail and make users retry if the error is transient.

If the error is not transient, that is super confusing and indicates a likely serious problem elsewhere in the system, either in its configuration or in our conversion code (?). I'm really of two minds here, I agree it's not a good user experience to fail at this particular point in the code, however failing here does get these serious errors the attention they deserve. (I consider conversion errors VERY serious as they potentially corrupt or invalidate persisted data.) Permitting these writes to proceed anyway escalates the error into permanent data loss (of the managed fields information). I'm not sure if that's ultimately better for users.

Yes, managed field information is less critical than other aspects of the object, but I expect that to be less true over time.

I would feel a lot better about dropping field managers if there was a plan for getting notified about the error that is the root cause.

cc @jpbetz agreed to look at the root causes a bit in the WG apply meeting.

**Comment by apelisse** [comment]
> cc @jpbetz agreed to look at the root causes a bit in the WG apply meeting.

To be super clear, the only one that we haven't investigated yet is:
https://github.com/openfaas/faas-netes/issues/630#issuecomment-625440515

It should be super easy to reproduce.

**Comment by liggitt** [comment]
Three distinct reported update failures related to managedFields indicates to me it is not yet reliable enough for field manager update errors to be made blocking for non-SSA write requests.

> Yes, managed field information is less critical than other aspects of the object, but I expect that to be less true over time.
>
> I would feel a lot better about dropping field managers if there was a plan for getting notified about the error that is the root cause.

I agree, but my immediate priority is to restore compatibility with existing API calls.

I would be in favor of a follow-up that adds metrics/logging for these "should not happen"-style errors, to prove we've resolved all known issues.

**Comment by lavalamp** [comment]
I haven't dug into the specific causes. Are they really SSA specific? Or is SSA just newly exercising things that are broken?

**Comment by apelisse** [comment]
Here's a possible alternative solution that will make you both semi-happy (or most likely both unhappy). 

We could fix the underlying issue in the next release (and cherry-pick it to older version) and disable errors in older versions. If we do that, we'll probably end-up finding exactly one bug per release and in a few years we'll maybe be bug free? :-)

But the good thing is that: we know 1.18 won't be broken anymore, and we have a chance to have the right solution, bug-free for 1.19.

**Comment by lavalamp** [comment]
Yeah, I can get behind that.

**Comment by apelisse** [comment]
@jpbetz who's looking at the errors

**Comment by liggitt** [comment]
restoring compatibility for <= 1.18 versions, fixing known issues for >= 1.19 sgtm

**Comment by apelisse** [comment]
OK So we definitely need to split things up, and I'd rather have changes properly isolated, even if it means we have a few cherry-picks to do. There are a few related problems here:
1. Clearing managedFIelds is not great today, and just broken for CRDs, we know how to fix that
2. We want to ignore errors tracking fields on updates (still fail for apply)
3. We want to fix the underlying issue, one is being investigated (openfaas), the other is the clearing for CRD in 1.

**Comment by apelisse** [comment]
Let's keep this open until we're happy with the various things we wanted to fix. Most of them are done, but we're not there yet.

**Comment by liggitt** [comment]
/priority important-soon

lowering priority since the create/update/patch paths are covered by a stop-gap

**Comment by liggitt** [comment]
This is resolved in 1.16.11+, 1.17.7+, 1.18.4+
